### ISSUE TITLE
how to check total percentage of no alignment after megalodon analysis done

### ISSUE THREAD

**Issue by emily288** [nanoporetech/megalodon#143]
Hello,

I'm trying to get modified base calls and I only have CPU version of guppy. Here is the script.
#Script
megalodon \
    /work/fast5 \
    --guppy-server-path /work/ont-guppy_4.4.1/ont-guppy/bin/guppy_basecall_server \
    --guppy-params "-d /work/rerio/basecall_models/" \
    --guppy-config res_dna_r941_min_modbases_5mC_CpG_v001.cfg \
    --outputs basecalls mappings mod_mappings mods \
    --reference /work/GRCm38.101/sequences.fa \
        --output-directory /work/megalodon_test \
    --mod-motif m CG 0 \
        --write-vcf-log-probs \
    --guppy-timeout 500 --reads-per-guppy-batch 50 --processes 8 --device 0 1 --overwrite > /workmegalodon_test/mega-bas.out

This is a screenshot of at the beginning of processing a run. 
![image](https://user-images.githubusercontent.com/83426502/118541456-b0063c80-b71f-11eb-975a-e5b6338ac107.png)
At the beginning, the percentage of guppy server timeout is 22%, it was around 6% after 2h running; however, the percentage of no alignment was still around 20% after 2h running.
Because I only have 2h to test my code on a remote cluster, or I need submit script to the cluster and get final results at the end. So my first question is: how to check the percentages of "guppy server timeout" and "no alignment" during the processing of a run and after a run has finished?
My second question is: how to modify my script to decrease the percentage of no alignment?

Thank you!

**Comment by marcus1487** [comment]
The processing result of each read is recorded in the megalodon log file. I would recommend parsing this file for lines containing these two errors in order to programmatically compute these proportions.

For advice on decreasing the percentage of reads producing no alignment, this is unfortunately quite a difficult question to answer. There are a multitude of root causes for this issue. Some simple things to check are that the sample is well matched to the provided reference (ideally providing a sample-specific polished reference, but this has issues when comparing to other samples), only provide "pass" FAST5 reads, filtering for adapter only reads, reduce contaminants in input, ensure basecalling model is matched to device/pore type. Note that this is only a small sample of the issues that can arise which would cause no alignment, but is a good start.

**Comment by emily288** [comment]
For the first question,

**Comment by emily288** [comment]
Thank you very much, that's helpful.
For the first question, just to make sure, did you mean this log.txt file (first picture), it still running... Once it's done, using the number of no alignment reads (or the number of guppy server timeout reads) divided by the total number of processed reads to get the error percentage, am I right? 
![image](https://user-images.githubusercontent.com/83426502/118553965-fca54400-b72e-11eb-8a43-0b61b6225da6.png)

And I know the lower error percentage number the better, but do we have a maximum value of error percentage to indicate that if the error percentage below this value we can tolerate it?

Thank you!

**Comment by marcus1487** [comment]
Yes. This is the log file. The exact computation is probably not possible (due to some weird edge case errors), for the most path the reads containing `[read_id] Failed` and `[read_id] Success` is a good measure of the processing results.

For the second question, this is very sample dependent. For some more experimental samples, 99.9% of reads failing may be a successful run since some reads were successfully processed and for some more standard samples 1% failed reads might be considered too high. Most samples will fall between this range, but again this is very dependent upon the sample and the goals for downstream analysis. I can't reliably give any more specific range than this unfortunately.

**Comment by emily288** [comment]
Got it! I really appreciate it.
But I don't understand what are edge case errors, could you please give an example? Do you mean I only need take [read_id] Failed and [read_id] Success (forth line in the previous picture) into account when calculate the percentage of errors?

For the second question, what are the "pass" fasr5 reads, what methods do you recommend to filter fast5 files? Because I only see two filtering options ( Qscore and  readlength) in this software we use to get fast5 files. 
![image](https://user-images.githubusercontent.com/83426502/118682079-1d70a680-b7ce-11eb-83d6-33446c9ceece.png)

Thank you!

**Comment by marcus1487** [comment]
There are several edge cases that are hit quite rarely, but can result in read failing for other reasons. For example if a read is mapped to a region with too many bases for the signal which produced the basecalls megalodon has no recourse to call modified bases for this read. This is a very extreme edge case that almost never happens, but do at some rate and cause the "failed" processing tag. There a number of other such checks that are made throughout processing.

In short though, yes just take the failed and success lines from the log file as shown in your previous screenshot.

For the second question, yes these are the options for filtering reads from minknow. In my previous comment about megalodon pass reads (between 1% and 99%) I am referring to reads that end in the "success" output type (not "no alignment" or another error).

**Comment by emily288** [comment]
I think I understand. Thank you very much!
